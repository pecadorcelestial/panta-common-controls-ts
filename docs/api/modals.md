# Modal

## Methods

| Name | Parameters | Type | Description |
|------|------------|------|-------------|
| show | N/A        | N/A  | Shows the notification... obviously. |
| hide | N/A        | N/A  | Hides the notification... also obvious. |

## Usage

```javascript
import React, { Component } from 'react';
import { Modal, ModalHeader, ModalTitle, ModalBody, ModalFooter } from 'panda-common-controls';

class BasicForm extends Component {
    render() {
        //Result.
        return(
            <div>
                <button onClick={() => { this.ModalRef.show(); }}>Show modal</button>
                <Modal ref={modal => this.ModalRef = modal}>
                    <ModalHeader>
                        <ModalTitle>Modal title</ModalTitle>
                    </ModalHeader>
                    <ModalBody>
                        <label>Body content.</label>
                    </ModalBody>
                    <ModalFooter>
                        <label>&#9400; Panda Corp. all rights reserved.</label>
                    </ModalFooter>
                </Modal>
            <div>
        );
    }
}
```
