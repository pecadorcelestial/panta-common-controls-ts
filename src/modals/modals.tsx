//Componentes generales.
import React, { HTMLAttributes, HTMLProps } from 'react';
import styled, { css, keyframes, Keyframes, StyledComponent, BaseThemedCssFunction } from 'styled-components';

//Componentes locales.
import { RoundButton } from '../buttons/buttons';

interface ILayoutProps {
	lock: boolean;
};
const Layout: StyledComponent<'div', any, ILayoutProps & HTMLProps<HTMLDivElement>, never> = styled.div`
	background-color: rgba(0,0,0,0.4);
    display: ${(props: ILayoutProps) => props.lock ? `block` : `none`};
	height: 100%;
	left: 0px;
	overflow: auto;
	position: fixed;
	top: 0px;
	width: 100%;
	z-index: 997;
`;

const Mask: StyledComponent<'div', any, HTMLProps<HTMLDivElement>, never> = styled.div`
	background-color: rgba(0,0,0,0.4);
    display: block;
	height: 100%;
	left: 0px;
	overflow: auto;
	position: fixed;
	top: 0px;
	width: 100%;
	z-index: 998;
`;

const show: Keyframes = keyframes`
	from {
		opacity: 0;
		top: -100%;
	}
  	to {
		opacity: 1;
		top: 0px;
	}
`;

const showWithBounce: Keyframes = keyframes`
	0% {
		top: -100%;
		opacity: 0;
	}
	80% {
		top: 50px;
		opacity: 0.8;
	}
	90% {
		top: 20px;
		opacity: 0.9;
	}
  	100% {
		top: 30px;
		opacity: 1;
	}
`;

const hide: Keyframes = keyframes`
	0% {
		opacity: 1;
		top: 0px;
	}
  	100% {
		opacity: 0;
		top: -100%;
	}
`;

const animationShow = css`
	${show} 0.6s ease-in-out forwards;
`;

const animationHide = css`
	${hide} 0.6s ease-in-out forwards;
`;

interface IContentProps extends HTMLProps<HTMLDivElement> {
	hide: boolean;
	show: boolean;
	showWithBounce: boolean;
};
const Content: StyledComponent<'div', any, IContentProps, never> = styled.div`
	background-color: #FFF;
    border-radius: 5px;
    box-shadow: 0 1px 8px rgba(0,0,0,0.05);	
    height: auto;
    margin: 50px auto;
    overflow: hidden;
	padding: 0px;
	position: relative;
	width: 600px;
	z-index: 999;
	
	animation: ${(props: IContentProps) => (props.show && !props.hide) ? animationShow : ((!props.show && props.hide) ? animationHide : 'unset')};
    
	@media screen and (max-width: 767px) {
		width: 90%;
	}
	
	@media screen and (min-width: 768px) and (max-width: 991px) {
		width: 90%;
	}
`;

export const ModalHeader: StyledComponent<'div', any, HTMLProps<HTMLDivElement>, never> = styled.div`
	background-color: #1476FB;
    box-sizing: border-box;
    display: inline-block;
    float: left;
    height: 45px;
	margin: 0px;
	padding: 10px;
    width: 100%;
`;

export const ModalTitle: StyledComponent<'label', any, HTMLProps<HTMLLabelElement>, never> = styled.label`
    box-sizing: border-box;
	color: #FFF;	
	height: 24px;
	font-family: "Open Sans", sans-serif;
	font-size: 18px;
	font-stretch: normal;
	font-style: normal;
	font-weight: bold;
	letter-spacing: normal;
	margin: 0px;
	padding: 3px 10px;
	text-align: left;
	width: 100%;
	
	@media screen and (max-width: 767px) {
		font-size: 12px;
		padding: 0px;
		width: calc(100% - 15px);
	}
	
	@media screen and (min-width: 768px) and (max-width: 991px) {
		font-size: 12px;
		padding: 0px;
		width: calc(100% - 15px);
	}
`;

export const ModalBody: StyledComponent<'div', any, HTMLProps<HTMLDivElement>, never> = styled.div`
    background-color: #FFF;
    box-sizing: border-box;
    display: inline-block;
    float: left;
    height: auto;
	margin: 0px;
	padding: 10px;
    width: 100%;
`;

export const ModalFooter: StyledComponent<'div', any, HTMLProps<HTMLDivElement>, never> = styled.div`
    background-color: #FFF;
    border-top: 1px solid #A7A7A7;
    box-sizing: border-box;
    display: inline-block;
    float: left;
    height: 40px;
	margin: 0px;
	padding: 10px;
    width: 100%;
`;

interface IModalProps extends HTMLAttributes<HTMLDivElement> {
	showWithBounce?: boolean;
};
interface IModalState {
	hide: boolean;
	lock: boolean;
	show: boolean;
};
export class Modal extends React.Component<IModalProps, IModalState> {
	//*** CONSTRUCTOR ***
	constructor(props: IModalProps) {
		super(props);
		this.state = {
            show: false,
            hide: false,
            lock: false
		};
	}
	//*** MÃ‰TODOS ***
	show = (): void => {
		this.setState({ show: true, hide: false, lock: true });
	}
	hide = (): void => {
		this.setState({ show: false, hide: true }, () => {
            setTimeout(() => { this.setState({ lock: false }); }, 600);
        });
	}
	//*** RESULTADOS ***
	render() {
		let { showWithBounce, children, ...rest } = this.props;
		return(
			<Layout lock={this.state.lock} {...rest}>
				<Content show={this.state.show} hide={this.state.hide} showWithBounce={showWithBounce || false}>
					{children}
					<RoundButton type='button' theme='flatWhite' size='small' icon='solidTimesCircle' style={{ position: 'absolute', right: '5px', top: '8px' }} onClick={(event: React.MouseEvent<HTMLButtonElement>) => { event.preventDefault(); this.hide(); }}/>
				</Content>
                {
                    this.state.lock ?
                    <Mask/> :
                    null
                }
			</Layout>
		);
	}
}