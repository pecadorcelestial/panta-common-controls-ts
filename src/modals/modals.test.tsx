
//RRRR  EEEEE  AAA   CCCC TTTTT
//R   R E     A   A C       T
//RRRR  EEE   AAAAA C       T
//R   R E     A   A C       T
//R   R EEEEE A   A  CCCC   T

import React from 'react';
import renderer from 'react-test-renderer';

//EEEEE N   N ZZZZZ Y   Y M   M EEEEE
//E     NN  N    Z   Y Y  MM MM E
//EEE   N N N   Z     Y   M M M EEE
//E     N  NN  Z      Y   M   M E
//EEEEE N   N ZZZZZ  YYY  M   M EEEEE

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//JJJJJ EEEEE  SSSS TTTTT
//    J E     S       T
//    J EEE    SSS    T
//J   J E         S   T
// JJJ  EEEEE SSSS    T

//NOTA: Éste módulo ayuda a que los errores en componentes estilizados se vean de manera correcta:
//https://medium.com/styled-components/effective-testing-for-styled-components-67982eb7d42b
import 'jest-enzyme';
import 'jest-styled-components';

// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT EEEEE  SSSS
//C     O   O MM MM P   P O   O NN  N E     NN  N   T   E     S
//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   EEE    SSS
//C     O   O M   M P     O   O N  NN E     N  NN   T   E         S
// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   EEEEE SSSS

import { Modal, ModalHeader, ModalTitle, ModalBody, ModalFooter } from './modals';

//Snapshot #1.
describe('[SNAPSHOT][Componentes][Common][Modal] - Modal (normal).', () => {
	//Tipo: Normal.
	it('Debe pintar el componente correctamente.', () => {
        //Se crea el componente.
        const modal = <Modal>
            <ModalHeader>
                <ModalTitle>Título del modal</ModalTitle>
            </ModalHeader>
            <ModalBody>
                <label>Cuerpo del modal</label>
            </ModalBody>
            <ModalFooter>
                <label>&#9400; Panda Corp. all rights reserved.</label>
            </ModalFooter>
        </Modal>;
        const component = renderer.create(modal).toJSON();
		expect(component).toMatchSnapshot();
	});
});

//Flujo + Métodos #1.
describe('[FLUJO][Componentes][Common][Modal] - Se muestra la ventana modal y se debe ocultar usando el botón dentro del mismo.', () => {
    //Se crea el componente.
    //NOTA: Al utilizar la función 'mount' se detona las siguientes funciones: constructor, componentDidMount y render.
    const modal = <Modal>
            <ModalHeader>
                <ModalTitle>Título del modal</ModalTitle>
            </ModalHeader>
            <ModalBody>
                <label>Cuerpo del modal</label>
            </ModalBody>
            <ModalFooter>
                <label>&#9400; Panda Corp. all rights reserved.</label>
            </ModalFooter>
        </Modal>;
    const component = enzyme.mount(modal);
    let button = component.find('button');
    it('Se revisa que exista 1 objeto tipo "button".', () => {
        expect(button).toHaveLength(1);
    });

    // SSSS H   H  OOO  W   W
    //S     H   H O   O W   W
    // SSS  HHHHH O   O W W W
    //    S H   H O   O WW WW
    //SSSS  H   H  OOO  W   W

	it('Debe mostrar la ventana modal.', () => {
		//Simulación.
		component.instance().show();
		//Expectativa.
        expect(component.state().show).toBe(true);
        expect(component.state().hide).toBe(false);
	});

    //H   H IIIII DDDD  EEEEE
    //H   H   I   D   D E
    //HHHHH   I   D   D EEE
    //H   H   I   D   D E
    //H   H IIIII DDDD  EEEEE

	//Se cambia el texto.
	it('Se espera a que la ventana modal se oculte la presionar el botón dentro.', () => {
        //Simulación.
        button.simulate('click', {});
        //Expectativa.
        expect(component.state().show).toBe(false);
        expect(component.state().hide).toBe(true);
	});

	//U   U N   N M   M  OOO  U   U N   N TTTTT
	//U   U NN  N MM MM O   O U   U NN  N   T
	//U   U N N N M M M O   O U   U N N N   T
	//U   U N  NN M   M O   O U   U N  NN   T
	// UUU  N   N M   M  OOO   UUU  N   N   T
    /*
    it('Se desmonta el componente.', () => {
        component.unmount();
    });
    */
});

//Flujo + Métodos #2.
describe('[FLUJO][Componentes][Common][Modal] - Se muestra la ventana modal y se debe ocultar usando el método expuesto.', () => {
    //Se crea el componente.
    //NOTA: Al utilizar la función 'mount' se detona las siguientes funciones: constructor, componentDidMount y render.
    const modal = <Modal>
            <ModalHeader>
                <ModalTitle>Título del modal</ModalTitle>
            </ModalHeader>
            <ModalBody>
                <label>Cuerpo del modal</label>
            </ModalBody>
            <ModalFooter>
                <label>&#9400; Panda Corp. all rights reserved.</label>
            </ModalFooter>
        </Modal>;
    const component = enzyme.mount(modal);
    let button = component.find('button');
    it('Se revisa que exista 1 objeto tipo "button".', () => {
        expect(button).toHaveLength(1);
    });

    // SSSS H   H  OOO  W   W
    //S     H   H O   O W   W
    // SSS  HHHHH O   O W W W
    //    S H   H O   O WW WW
    //SSSS  H   H  OOO  W   W

	it('Debe mostrar la notificación.', () => {
		//Simulación.
		component.instance().show();
		//Expectativa.
        expect(component.state().show).toBe(true);
        expect(component.state().hide).toBe(false);
	});

    //H   H IIIII DDDD  EEEEE
    //H   H   I   D   D E
    //HHHHH   I   D   D EEE
    //H   H   I   D   D E
    //H   H IIIII DDDD  EEEEE

	//Se cambia el texto.
	it('Se espera a que la notificación se oculte al hacer la llamada al método expuesto.', (done) => {
        //Simulación.
        component.instance().hide();
        //Expectativa.
        expect(component.state().show).toBe(false);
        expect(component.state().hide).toBe(true);
        setTimeout(() => {
            done();
            expect(component.state().lock).toBe(false);
        }, 700);
	});

	//U   U N   N M   M  OOO  U   U N   N TTTTT
	//U   U NN  N MM MM O   O U   U NN  N   T
	//U   U N N N M M M O   O U   U N N N   T
	//U   U N  NN M   M O   O U   U N  NN   T
	// UUU  N   N M   M  OOO   UUU  N   N   T
    
    it('Se desmonta el componente.', () => {
        component.unmount();
    });    
});
