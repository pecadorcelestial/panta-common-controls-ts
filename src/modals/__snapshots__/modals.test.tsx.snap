// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT][Componentes][Common][Modal] - Modal (normal). Debe pintar el componente correctamente. 1`] = `
.c9 {
  display: inline-block;
  fill: #FFF;
  height: 50px;
  margin: 0px;
  vertical-align: top;
  width: 50px;
  -webkit-transition: all .3s;
  transition: all .3s;
}

.c7 {
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 5px;
  box-sizing: border-box;
  color: #FFF;
  cursor: pointer;
  display: inline-block;
  font-size: 14px;
  font-family: "Open Sans",sans-serif;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  height: 30px;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 20px;
  padding: 5px 30px;
  text-align: center;
  width: auto;
  -webkit-transition: all .3s;
  transition: all .3s;
}

.c7:hover {
  background-color: rgba(235,235,235,0.5);
  border: 1px solid rgba(235,235,235,0);
  color: #FFF;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7:hover .c8 {
  fill: #FFF;
}

.c7:disabled {
  background-color: transparent;
  border: 1px solid transparent;
  color: #BFBFBF;
  pointer-events: none;
}

.c7:disabled .c8 {
  fill: #BFBFBF;
}

.c7:focus {
  outline: none;
}

.c7 .c8 {
  fill: #FFF;
  height: 18px;
  width: 18px;
}

.c6 {
  border-radius: 50%;
  padding: 1px 0px 0px 0px;
  width: 30px;
}

.c0 {
  background-color: rgba(0,0,0,0.4);
  display: none;
  height: 100%;
  left: 0px;
  overflow: auto;
  position: fixed;
  top: 0px;
  width: 100%;
  z-index: 997;
}

.c1 {
  background-color: #FFF;
  border-radius: 5px;
  box-shadow: 0 1px 8px rgba(0,0,0,0.05);
  height: auto;
  margin: 50px auto;
  overflow: hidden;
  padding: 0px;
  position: relative;
  width: 600px;
  z-index: 999;
  -webkit-animation: unset;
  animation: unset;
}

.c2 {
  background-color: #1476FB;
  box-sizing: border-box;
  display: inline-block;
  float: left;
  height: 45px;
  margin: 0px;
  padding: 10px;
  width: 100%;
}

.c3 {
  box-sizing: border-box;
  color: #FFF;
  height: 24px;
  font-family: "Open Sans",sans-serif;
  font-size: 18px;
  font-stretch: normal;
  font-style: normal;
  font-weight: bold;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0px;
  padding: 3px 10px;
  text-align: left;
  width: 100%;
}

.c4 {
  background-color: #FFF;
  box-sizing: border-box;
  display: inline-block;
  float: left;
  height: auto;
  margin: 0px;
  padding: 10px;
  width: 100%;
}

.c5 {
  background-color: #FFF;
  border-top: 1px solid #A7A7A7;
  box-sizing: border-box;
  display: inline-block;
  float: left;
  height: 40px;
  margin: 0px;
  padding: 10px;
  width: 100%;
}

@media screen and (max-width:767px) {
  .c1 {
    width: 90%;
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c1 {
    width: 90%;
  }
}

@media screen and (max-width:767px) {
  .c3 {
    font-size: 12px;
    padding: 0px;
    width: calc(100% - 15px);
  }
}

@media screen and (min-width:768px) and (max-width:991px) {
  .c3 {
    font-size: 12px;
    padding: 0px;
    width: calc(100% - 15px);
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <label
        className="c3"
      >
        Título del modal
      </label>
    </div>
    <div
      className="c4"
    >
      <label>
        Cuerpo del modal
      </label>
    </div>
    <div
      className="c5"
    >
      <label>
        Ⓒ Panda Corp. all rights reserved.
      </label>
    </div>
    <button
      className="c6 c7"
      icon="solidTimesCircle"
      onClick={[Function]}
      size="small"
      style={
        Object {
          "position": "absolute",
          "right": "5px",
          "top": "8px",
        }
      }
      theme="flatWhite"
      type="button"
    >
      <svg
        alt="solidTimesCircle"
        className="c8 c9"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<path d=\\"M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm121.6 313.1c4.7 4.7 4.7 12.3 0 17L338 377.6c-4.7 4.7-12.3 4.7-17 0L256 312l-65.1 65.6c-4.7 4.7-12.3 4.7-17 0L134.4 338c-4.7-4.7-4.7-12.3 0-17l65.6-65-65.6-65.1c-4.7-4.7-4.7-12.3 0-17l39.6-39.6c4.7-4.7 12.3-4.7 17 0l65 65.7 65.1-65.6c4.7-4.7 12.3-4.7 17 0l39.6 39.6c4.7 4.7 4.7 12.3 0 17L312 256l65.6 65.1z\\"/>",
          }
        }
        viewBox="0 0 512 512"
        x="0px"
        y="0px"
      />
    </button>
  </div>
</div>
`;
